### DISCUSSION TITLE
Can the BaseAddress be changed after the service is injected?

### DISCUSSION THREAD

**Discussion by michaelakin** [reactiveui/refit#1410] [Q&A] [answered]
Hi, I would like to change the base address on particular api endpoints that have been injected with a refit service.  Is this posssilbe, or can I only do it when Getting a required service?

```
app.MapGet("/getip", async ([FromServices] IIpAddressApi ipApi) =>
{
    // would like to change the base address here for services like this based on something.
    Log.Information("Getting the ip address");
    var result = await ipApi.GetIp();
    return result;
}).WithName("getip");

```

here  is the set up for that service:

```
builder.Services.AddRefitClient<IIpAddressApi>()
        .ConfigureHttpClient(c => c.BaseAddress = new Uri("https://ipapi.co"))
        .AddTransientHttpErrorPolicy(x =>
        x.WaitAndRetryAsync(
            Backoff.DecorrelatedJitterBackoffV2(medianFirstRetryDelay: TimeSpan.FromSeconds(2),
            retryCount: 5), onRetryAsync: async (outcome, span, retryAttempt, context) =>
            {
                Log.Logger.Warning("Delaying for {delay}ms, then making retry {retry}.", span.TotalMilliseconds, retryAttempt);
            }
        ));
```

and Here is the service:

```
using Refit;
using Our.Enterprise..Models;

namespace Our.Enterprise.Interfaces
{
    internal interface IIpAddressApi
    {
        [Headers("User-Agent: NotChrome")]
        [Get("/json")]
        public Task<IpDto> GetIp();
    }
}
```

**Comment by EmilAlipiev** [comment]
Have you ever found an answer for this?  I only can get it working like below but this not overriding, changing, rather getting a new instance with a new url.
```
  private IApi GetClient()
    {
            return ClientFactory.CreateClient(apiUrl);
    }
```

**Comment by michaelakin** [comment]
I never got an answer and I don't remember if I could do it after injection , but here is how I did it:

Setup:
```
builder.Services.AddHttpClient("payment")
.AddTransientHttpErrorPolicy(x =>
x.WaitAndRetryAsync(
    Backoff.DecorrelatedJitterBackoffV2(medianFirstRetryDelay: TimeSpan.FromSeconds(1),
    retryCount: 5), onRetryAsync: async (outcome, span, retryAttempt, context) =>
    {
        Log.Logger.Warning("Delaying for {delay}ms, then making retry {retry}.", span.TotalMilliseconds, retryAttempt);
    }
))
.AddHttpMessageHandler<HttpLoggingHandler>()
.AddTypedClient<IProxyService, ProxyService>();
```

Injection of Http Client:
```
public ProxyService(
    ILogger<ProxyService> logger,
    HttpClient httpClient)
{
   _logger = logger;
   _httpClient = httpClient;

}
```

use in a method to set the base address
```
var auth = await GetAuthInfoAsync(apiGatewayStage, "payment");
if (_httpClient.BaseAddress == null)
{
    _httpClient.BaseAddress = new Uri(auth.BaseUrl);
}
var paymentApi = RestService.For<IPaymentApi>(_httpClient, GetRefitSettings());
```

Hopefully this helps.

**Comment by michaelakin** [comment]
I find it interesting because the generated code exposes a client where you should be able to change the base address, but it is not available for use

```
    partial class EnterpriseInterfacesIIpAddressApi
        : global::Enterprise.Interfaces.IIpAddressApi

    {
        /// <inheritdoc />
       public global::System.Net.Http.HttpClient Client { get; }
        readonly global::Refit.IRequestBuilder requestBuilder;

        /// <inheritdoc />
        public EnterpriseInterfacesIIpAddressApi(global::System.Net.Http.HttpClient client, global::Refit.IRequestBuilder requestBuilder)
        {
            Client = client;
            this.requestBuilder = requestBuilder;
        }
```

I would like to find out more why that is not available.

![image](https://github.com/reactiveui/refit/assets/7062652/fba5dce4-2714-49f2-856b-c7e500c00318)

**Comment by glennawatson** [comment, marked as answer]
Short version: do not try to mutate the base address on the injected typed client. With AddRefitClient<T> the HttpClient is created and configured per call by IHttpClientFactory, so reaching in to change BaseAddress is fragile and is not exposed on the generated interface by design.

Pick one of these depending on the need:

1. The address depends on runtime data per request -> use a DelegatingHandler that rewrites the request URI. This is the recommended pattern and works cleanly with DI and Polly:

```csharp
class BaseAddressHandler(IBaseUrlProvider provider) : DelegatingHandler
{
    protected override Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request, CancellationToken ct)
    {
        var baseUri = new Uri(provider.GetCurrentBaseUrl());
        request.RequestUri = new Uri(baseUri, request.RequestUri!.PathAndQuery);
        return base.SendAsync(request, ct);
    }
}

builder.Services.AddRefitClient<IIpAddressApi>()
    .ConfigureHttpClient(c => c.BaseAddress = new Uri("https://placeholder"))
    .AddHttpMessageHandler<BaseAddressHandler>();
```

ConfigureHttpClient still needs to set some BaseAddress (Refit requires one), but the handler overrides the host per request.

2. A fixed set of distinct endpoints -> register a separate typed client per endpoint (different interface or different named registration), each with its own ConfigureHttpClient base address, and inject the one you need.

3. The address is only known at call time and you are not using DI -> build the client yourself with RestService.For<T>(httpClient) where you set httpClient.BaseAddress, as one of the earlier replies showed.

On the generated Client property being read-only: that is intentional. The instance is bound to the HttpClient handed to it at construction; the supported extension point for per-request routing is a DelegatingHandler, not swapping BaseAddress on a shared injected client (which would race across concurrent callers).
